content/blog/2019-09-25-bril-to-llvm-ocaml.md

```diff
@@ -0,0 +1,151 @@
++++
+title = "Bril to LLVM using OCaml"
+extra.author = "Daniel Weber"
+extra.bio = """
+  [Daniel Weber](https://github.com/Dan12) is an MEng student interested in programming languages, distributed systems, and the inner workings of computers.
+"""
++++
+
+## The Goal
+
+For this project, I wanted to create a transformation from Bril to LLVM IR and implement this transformation in OCaml. The motivation for the first goal of the project (LLVM code generation) was to allow Bril to be compiled and run natively, instead of just interpreted. LLVM can be compiled to machine code using a tool like `clang`. Furthermore, LLVM IR supports many optimizations which allow even a naive transformation of Bril to LLVM to be quite performant. The motivation for the second goal was that OCaml provides several constructs, like variants, GADTs, and partial function application. These features might make it nicer to write IR transformations than it would to do the transformations in Typescript.
```

Comment:
> LLVM can be compiled to machine code using a tool like `clang`.

This is a slightly confusing way to put it because the Clang project is *not* about compiling LLVM to machine code—it's about compiling C, C++, and Objective-C to LLVM IR. So maybe just say something like "LLVM can produce machine code" or something?

Reply:
Hmm, how about just deleting
"LLVM can be compiled to machine code using a tool like `clang`."